java/ql/lib/semmle/code/java/security/WebviewDubuggingEnabledQuery.qll

```diff
@@ -0,0 +1,41 @@
+/** Definitions for the Android Webview Debugging Enabled query */
+
+import java
+import semmle.code.java.dataflow.DataFlow
+import semmle.code.java.controlflow.Guards
+import semmle.code.java.security.SecurityTests
+
+/** Holds if `ex` looks like a check that this is a debug build. */
+private predicate isDebugCheck(Expr ex) {
+  exists(Expr subex, string debug |
+    debug.toLowerCase().matches(["%debug%", "%test%"]) and
+    subex.getParent*() = ex
+  |
+    subex.(VarAccess).getVariable().getName() = debug
+    or
+    subex.(MethodAccess).getMethod().hasName("getProperty") and
+    subex.(MethodAccess).getAnArgument().(CompileTimeConstantExpr).getStringValue() = debug
+  )
+}
+
+/** A configuration to find instances of `setWebContentDebuggingEnabled` called with `true` values. */
+class WebviewDebugEnabledConfig extends DataFlow::Configuration {
+  WebviewDebugEnabledConfig() { this = "WebviewDebugEnabledConfig" }
+
+  override predicate isSource(DataFlow::Node node) {
+    node.asExpr().(BooleanLiteral).getBooleanValue() = true
+  }
+
+  override predicate isSink(DataFlow::Node node) {
+    exists(MethodAccess ma |
+      ma.getMethod().hasQualifiedName("android.webkit", "WebView", "setWebContentsDebuggingEnabled") and
+      node.asExpr() = ma.getArgument(0)
+    )
+  }
+
+  override predicate isBarrier(DataFlow::Node node) {
+    exists(Guard debug | isDebugCheck(debug) and debug.controls(node.asExpr().getBasicBlock(), _))
```

Comment:
We are sanitizing both branches here, and I think that might be a little too much. Is this done this way to handle the `0 != (getApplicationInfo().flags & ApplicationInfo.FLAG_DEBUGGABLE` case?

If that's the case, would it make sense to implement more precise guards? Namely, something like this: https://github.com/github/codeql/pull/10177/files#diff-38978229f15815dd860ec9f3696f469770f85b8082736b808072ef64be18a834R202

So if the `debug` variable is of type boolean, we can assume it sanitizes the branch `true`, whereas if it's an `int` (or anything else, really), we can expect there'll be some `EqualityTest` involved and then sanitize the branch indicated by `EqualityTest::polarity`. What do you think?

Reply:
How much benefit is there to making the guard more precise? It doesn't seem that likely to me that somebody would write a check for a debug flag but then enable debug features in the wrong branch. And there are different possible variations; e.g. an integer might be checked that it's >0 instead, which we might not catch by only looking for equality tests.

Reply:
I ran a query to find out how many sinks were sanitized by this guard in 1k projects, and only 4 results came back. Those 4 results were under the correct branch, so I think your assumption makes sense and we don't need to increase the precision of the guard. We can always go back and do it if we spot many FNs caused by this in the future.